Sources/NIO/SocketChannel.swift

```diff
@@ -294,11 +294,17 @@ class BaseSocketChannel<T: BaseSocket>: SelectableChannel, ChannelCore {
             recvAllocator = value as! RecvByteBufferAllocator
         case _ as AutoReadOption:
             let auto = value as! Bool
-            autoRead = auto
-            if auto {
-                read0()
-            } else {
-                pauseRead0()
+            let old = self.autoRead
+            self.autoRead = auto
+
+            // We only want to call read0() or pauseRead0() if we already registered to the EventLoop if not this will be automatically done
+            // once register0 is called. Beside this we also only need to do it when the value actually change.
+            if !neverRegistered && old != auto {
```

Comment:
This raises a question; what should happen if you set a channel option on a channel that *was* registered but is now closed? Presumably not an assertion error, right?

Reply:
@Lukasa I guess we should fail the `EventLoopFuture`. Should I do this as part of this PR or do a follow up + tests ?

Reply:
@normanmaurer / @Lukasa I think what Norman proposes makes sense. Sounds like a follow-up that needs a test to me

Reply:
If @Lukasa agrees I will do this after this one is merged...

Reply:
Agreed, follow-up with tests sounds fine.

Reply:
@weissi @Lukasa here you go https://github.com/apple/swift-nio/pull/198